src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1078,27 +1078,36 @@
                            fact-binding (conj fact-binding))
 
             ;; Find children that all parent nodes have.
-            common-children-ids (into
-                                 ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
-                                 ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
-                                 ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
-                                 ;; Having different nodes under the same
-                                 ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
-                                 (sorted-set)
-                                 (apply s/union (-> (:forward-edges beta-graph)
-                                                    (select-keys parent-ids)
-                                                    (vals))))
-
+            forward-edges (if (= 1 (count parent-ids))
+                            ;; If there is only one parent then there is no need to reconstruct
+                            ;; the set of forward edges
+                            (-> (:forward-edges beta-graph)
+                                (get (first parent-ids)))
+                            (->> (select-keys parent-ids (:forward-edges beta-graph))
+                                 vals
+                                 (into [] (comp cat (distinct)))))
```

Comment:
We could probably improve on Clojure'd distinct transducer here for the case of integers, since it just [uses a persistent set](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core.clj#L4976), and we could use either a transient set or a logical set backed by an array with a length set to prevent overflow.  IMO that would be premature evidence until/unless we have evidence that this is a perf problem.

Overall I expect removing the sorted-set here to be a significant benefit.  From profiling it seems it performs really poorly, which is not entirely surprising since its implementation seems to build iteratively rather than using some kind of transient strategy, presorting when provided all elements at once, etc.